proc-macros/src/new/render_server.rs

```diff
@@ -0,0 +1,234 @@
+			// Add `SubscriptionSink` as the second input parameter to the signature.
+			let subscription_sink: syn::FnArg = syn::parse_quote!(subscription_sink: #subscription_sink_ty);
+			sub.signature.sig.inputs.insert(1, subscription_sink);
+			sub.signature
+		});
+
+		Ok(quote! {
+			#(#methods)*
+			#(#subscriptions)*
+		})
+	}
+
+	fn render_into_rpc(&self) -> Result<TokenStream2, syn::Error> {
+		let jrps_error = self.jrps_server_item(quote! { error::Error });
+		let rpc_module = self.jrps_server_item(quote! { RpcModule });
+
+		let methods = self.methods.iter().map(|method| {
+			// Rust method to invoke (e.g. `self.<foo>(...)`).
+			let rust_method_name = &method.signature.sig.ident;
+			// Name of the RPC method (e.g. `foo_makeSpam`).
+			let rpc_method_name = self.rpc_identifier(&method.name);
+			// `parsing` is the code associated with parsing structure from the
+			// provided `RpcParams` object.
+			// `params_seq` is the comma-delimited sequence of parametsrs.
+			let (parsing, params_seq) = self.render_params_decoding(&method.params);
+
+			if method.signature.sig.asyncness.is_some() {
+				quote! {
+					rpc.register_async_method(#rpc_method_name, |params, context| {
+						let owned_params = params.owned();
+						let fut = async move {
+							let params = owned_params.borrowed();
+							#parsing
+							Ok(context.as_ref().#rust_method_name(#params_seq).await)
+						};
+						Box::pin(fut)
+					})?;
+				}
+			} else {
+				quote! {
+					rpc.register_method(#rpc_method_name, |params, context| {
+						#parsing
+						Ok(context.#rust_method_name(#params_seq))
+					})?;
+				}
+			}
+		});
+
+		let subscriptions = self.subscriptions.iter().map(|sub| {
+			// Rust method to invoke (e.g. `self.<foo>(...)`).
+			let rust_method_name = &sub.signature.sig.ident;
+			// Name of the RPC method to subscribe (e.g. `foo_sub`).
+			let rpc_sub_name = self.rpc_identifier(&sub.name);
+			// Name of the RPC method to unsubscribe (e.g. `foo_sub`).
+			let rpc_unsub_name = self.rpc_identifier(&sub.unsub_method);
+			// `parsing` is the code associated with parsing structure from the
+			// provided `RpcParams` object.
+			// `params_seq` is the comma-delimited sequence of parametsrs.
+			let (parsing, params_seq) = self.render_params_decoding(&sub.params);
+
+			quote! {
+				rpc.register_subscription(#rpc_sub_name, #rpc_unsub_name, |params, sink, context| {
+					#parsing
+					Ok(context.as_ref().#rust_method_name(sink, #params_seq))
+				})?;
+			}
+		});
+
+		let doc_comment = "Collects all the methods and subscriptions defined in the trait \
+								and adds them into a single `RpcModule`.";
+
+		Ok(quote! {
+			#[doc = #doc_comment]
+			fn into_rpc(self) -> Result<#rpc_module<Self>, #jrps_error> {
+				let mut rpc = #rpc_module::new(self);
+
+				#(#methods)*
+				#(#subscriptions)*
+
+				Ok(rpc)
+			}
+		})
+	}
+
+	fn render_params_decoding(&self, params: &[(syn::PatIdent, syn::Type)]) -> (TokenStream2, TokenStream2) {
```

Comment:
I believe this allocates when using `serde_json::from_value` whereas [RpcParams doesn't](https://github.com/paritytech/jsonrpsee/blob/master/types/src/v2/params.rs#L63), 

I think our zero-alloc friend @maciejhirsz isn't super happy about that but we could take care of that in another pr :man_shrugging:

Reply:
*I believe* that for the first iteration it's not that important to be fully optimal here. Generic non-allocating implementation will be **much** less trivial to implement, and if the performance is the key, there always a way to optimize things manually.

Generally speaking, by writing the code yourself you are more likely to get better performance.